Source/modules/webcl/WebCLProgram.cpp

```diff
@@ -0,0 +1,502 @@
+    WebCLProgram* program = nullptr;
+    if (callback) {
+        if (m_callback) {
+            es.throwWebCLException(WebCLException::INVALID_OPERATION, WebCLException::invalidOperationMessage);
+            return;
+        }
+
+        // Store the callback, eventList to HashTable and call callbackProxy.
+        m_callback = callback;
+        callbackProxyPtr = &callbackProxy;
+        program = this;
+    }
+
+    cl_int err = CL_SUCCESS;
+    Vector<cl_device_id> clDevices;
+    Vector<RefPtr<WebCLDevice>> contextDevices = context()->getDevices();
+    if (devices.size()) {
+        Vector<cl_device_id> inputDevices;
+        for (auto device : devices)
+            inputDevices.append(device->getDeviceId());
+
+        size_t contextDevicesLength = contextDevices.size();
+        for (size_t z, i = 0; i < inputDevices.size(); i++) {
+            // Check if the inputDevices[i] is part of programs WebCLContext.
+            for (z = 0; z < contextDevicesLength; z++) {
+                if (contextDevices[z]->getDeviceId() == inputDevices[i]) {
+                    break;
+                }
+            }
+
+            if (z == contextDevicesLength) {
+                es.throwWebCLException(WebCLException::INVALID_DEVICE, WebCLException::invalidDeviceMessage);
+                return;
+            }
+
+            clDevices.append(inputDevices[i]);
+        }
+    } else {
+        for (auto contextDevice : contextDevices)
+            clDevices.append(contextDevice->getDeviceId());
+    }
+
+    if (!clDevices.size()) {
+        es.throwWebCLException(WebCLException::INVALID_VALUE, WebCLException::invalidValueMessage);
+        return;
+    }
+
+    m_isProgramBuilt = true;
+    err = clBuildProgram(m_clProgram, clDevices.size(), clDevices.data(), buildOptions.utf8().data(), callbackProxyPtr, program);
+
+    if (err != CL_SUCCESS)
+        es.throwWebCLException(WebCLException::BUILD_PROGRAM_FAILURE, WebCLException::buildProgramFailureMessage);
+}
+
+void WebCLProgram::build(const String& options, WebCLCallback* callback, ExceptionState& es)
+{
+    Vector<RefPtr<WebCLDevice>> devices;
+    build(devices, options, callback, es);
+}
+
+void WebCLProgram::release()
+{
+    if (isReleased())
+        return;
+
+    cl_int err = clReleaseProgram(m_clProgram);
+    if (err != CL_SUCCESS)
+        return;
+
+    m_clProgram = 0;
+}
+
+const String& WebCLProgram::sourceWithCommentsStripped()
+{
+    if (m_programSourceWithCommentsStripped.isNull())
+        removeComments(m_programSource, m_programSourceWithCommentsStripped);
+
+    return m_programSourceWithCommentsStripped;
+}
+
+WebCLProgram::WebCLProgram(cl_program program, WebCLContext* context, const String& kernelSource)
+    : WebCLObject(context)
+    , m_callback(nullptr)
+    , m_programSource(kernelSource)
+    , m_isProgramBuilt(false)
+    , m_clProgram(program)
+{
+}
+
+void WebCLProgram::callbackProxy(cl_program program, void* userData)
+{
```

Comment:
It's better to do nothing but call callbackProxyOnMainThread()
It makes us not use mutex and more solid to thread bug.
callbackProxyOnMainThread() should have all logic.